test/modeling_library/product.jl

```diff
@@ -0,0 +1,89 @@
+discrete_product = ProductDistribution(bernoulli, binom)
+
+@testset "product of discrete distributions" begin
+    @test is_discrete(discrete_product)
+    grad_bools = (has_output_grad(discrete_product), has_argument_grads(discrete_product)...)
+    @test grad_bools == (false, true, false, true)
+
+    p1 = 0.5
+    (n, p2) = (3, 0.9)
+
+    # random
+    x = unrelated_pair_product(p1, n, p2)
```

Comment:
We should add a type check here to ensure that `x` is a tuple of the appropriate type.
In particular, it should be the case that `typeof(x) == get_return_type(discrete_product)`. I would also add a check that `get_return_type(discrete_product)` returns the expected tuple type.

We should also check that the sampled numbers are within their respective supports (e.g. either true or false, or a positive integer).